Fix SELECT over DML queries

InnerLife0 opened the pull request:
Previously, using of SELECT over DML operations caused various errors.
___________
The first block of errors is related to distributed replicated tables. There was no motion node between SELECT and DML plan nodes. This caused SELECT to work with empty result set, no matter how many rows were processed by DML beneath.
Another error was with joining such result set with any table. There were assertion errors due to incompatibility of CdbLocusType_Replicated with SELECT statements.
The fix adds additional analysis of hasModifyingCTE flag which assumes SELECT query can have DML beneath.
___________
The second block of errors is related to quals pushed down from SELECT to DML operation. It's incorrect to push down quals to DML operations, because manipulated dataset can be filtered by wrong upper quals. It also caused indistinct rewrite error.
___________
The third block of errors is related to multiple references to CTE in SELECT. It causes duplicated DML operations or mutation errors during apply_motion() of cdbparallelize().
To fix this, we restrict SELECT queries with multiple references to CTE.

seryozhasmirnov commented on src/test/regress/expected/with_clause.out:
```diff
@@ -2180,3 +2180,174 @@ WITH RECURSIVE r1 AS (
+
+-- Test select over update queries. Check no qual pushed down to update from
+-- select, because it's incorrect to push it down in such way. Queries should
+-- be correctly executed without any rewriting error, e.g. rewriteManip.c:1409.
```
Probably it is better  to specify log message as code at [rewriteManip.c:1409](https://github.com/arenadata/gpdb/blob/2aea19802afbac1d0fa2f4b7c7d7eec5de9e3eaa/src/backend/rewrite/rewriteManip.c#L1409) can change its number later.

> "could not find replacement targetlist entry for attno %d"

InnerLife0 replied:
Everything can change. Code line, error message, assert condition, function name. If somebody will really need to find why test was created, he could easily switch to commit, where the test initially appeared, and so, find the exact and accurate error line by copypasting `rewriteManip.c:1409`. Searching by text (even if it was not changed) is more complex, because the error message can be duplicated in the multiple parts of code. As opposed to this, the line number is a direct link.

maksm90 replied:
> Probably it is better to specify log message as code at rewriteManip.c:1409 can change its number later.

+1. It might be not so convenient to jump to code line (that might not correspond to the target one over time) to realize the error that is covered by current test case

seryozhasmirnov commented on src/test/regress/expected/with_clause_optimizer.out:
```diff
@@ -2179,3 +2179,174 @@ WITH RECURSIVE r1 AS (
+
+-- Test select over update queries. Check no qual pushed down to update from
+-- select, because it's incorrect to push it down in such way. Queries should
+-- be correctly executed without any rewriting error, e.g. rewriteManip.c:1409.
```
Probably it is better to specify log message as code at [rewriteManip.c:1409](https://github.com/arenadata/gpdb/blob/2aea19802afbac1d0fa2f4b7c7d7eec5de9e3eaa/src/backend/rewrite/rewriteManip.c#L1409) can change its number later.

> "could not find replacement targetlist entry for attno %d"

> ERROR:  could not find replacement targetlist entry for attno 1 (rewriteManip.c:1410)

seryozhasmirnov commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -358,3 +358,63 @@ WITH RECURSIVE r1 AS (
+) select count(*) from cte;
+-- Test select over update queries. Check no qual pushed down to update from
+-- select, because it's incorrect to push it down in such way. Queries should
+-- be correctly executed without any rewriting error, e.g. rewriteManip.c:1409.
```
Probably it is better to specify log message as code at [rewriteManip.c:1409](https://github.com/arenadata/gpdb/blob/2aea19802afbac1d0fa2f4b7c7d7eec5de9e3eaa/src/backend/rewrite/rewriteManip.c#L1409) can change its number later.

> "could not find replacement targetlist entry for attno %d"

> ERROR:  could not find replacement targetlist entry for attno 1 (rewriteManip.c:1410)

seryozhasmirnov commented on src/test/regress/expected/with_clause.out:
```diff
@@ -2180,3 +2180,174 @@ WITH RECURSIVE r1 AS (
+(1 row)
+
+-- Test select over delete queries. Test joining, which caused fatal error
+-- (cdbpath.c:1105) due to incompatibility of using CdbLocusType_Replicated
```
Please specify what is the fatal error - i.e. failed assertion at [cdbpath.c:1105](https://github.com/arenadata/gpdb/blob/2aea19802afbac1d0fa2f4b7c7d7eec5de9e3eaa/src/backend/cdb/cdbpath.c#L1105) as line number can change later:

` Assert(root->upd_del_replicated_table > 0);`

seryozhasmirnov commented on src/test/regress/expected/with_clause_optimizer.out:
```diff
@@ -2179,3 +2179,174 @@ WITH RECURSIVE r1 AS (
+(1 row)
+
+-- Test select over delete queries. Test joining, which caused fatal error
+-- (cdbpath.c:1105) due to incompatibility of using CdbLocusType_Replicated
```
Please specify what is the fatal error - i.e. failed assertion at [cdbpath.c:1105](https://github.com/arenadata/gpdb/blob/2aea19802afbac1d0fa2f4b7c7d7eec5de9e3eaa/src/backend/cdb/cdbpath.c#L1105) as line number can change later:

` Assert(root->upd_del_replicated_table > 0);`

seryozhasmirnov commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -358,3 +358,63 @@ WITH RECURSIVE r1 AS (
+    update with_dml_dr set j = j + 1 where i <= 5 returning j
+) select count(*) from cte where j > 100;
+-- Test select over delete queries. Test joining, which caused fatal error
+-- (cdbpath.c:1105) due to incompatibility of using CdbLocusType_Replicated
```
Please specify what is the fatal error - i.e. failed assertion at [cdbpath.c:1105](https://github.com/arenadata/gpdb/blob/2aea19802afbac1d0fa2f4b7c7d7eec5de9e3eaa/src/backend/cdb/cdbpath.c#L1105) as line number can change later:

` Assert(root->upd_del_replicated_table > 0);`

Stolb27 commented:
@maksm90 please take a look at this one

maksm90 commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -2012,7 +2012,12 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
-		if (!cte->cterecursive)
+		/*
+		 * Additionally to recursive CTE queries, don't try to push down quals
+		 * to non-SELECT queries. There can be DML operation with RETURNING
```
Here and further: acronym "DML" in some contexts also incorporates SELECT statements. Therefore, it's better to specify operation as "modifying DML".

InnerLife0 replied:
I think Postgres [docs](https://www.postgresql.org/docs/9.4/dml.html) are pretty clean about what data manipulation operations are. Thanks.

maksm90 commented on src/backend/optimizer/plan/planner.c:
```diff
@@ -914,6 +914,23 @@ subquery_planner(PlannerGlobal *glob, Query *parse,
+			 * There is a bug in SELECT over DML operations.
+			 * CdbLocusType_Replicated copied to upper (related to SELECT)
+			 * nodes causing conflicts in multiple places, e.g. in
+			 * cdbpath_motion_for_join(), as such locus type is unxpected for
```
*unexpected

InnerLife0 replied:
Thanks.

maksm90 commented on src/backend/optimizer/plan/planner.c:
```diff
@@ -914,6 +914,23 @@ subquery_planner(PlannerGlobal *glob, Query *parse,
+			 * nodes causing conflicts in multiple places, e.g. in
+			 * cdbpath_motion_for_join(), as such locus type is unxpected for
+			 * SELECT queries. We force motion to avoid this.
+			 * 
```
There is a trailing space

InnerLife0 replied:
Thanks.

maksm90 commented on src/backend/optimizer/plan/setrefs.c:
```diff
@@ -1380,6 +1380,24 @@ trivial_subqueryscan(SubqueryScan *plan)
 			   *lc;
+	Movement	m_subplan,
+				m_scanplan;
+	
```
Trailing spaces

InnerLife0 replied:
Thanks.

maksm90 commented on src/backend/optimizer/plan/setrefs.c:
```diff
@@ -1380,6 +1380,24 @@ trivial_subqueryscan(SubqueryScan *plan)
+			plan->scan.plan.flow->req_move :
+			MOVEMENT_NONE;
+	/*
+	 * GPDB: Stripping out of SubqueryScan node causing overwriting of
```
*causes ?

InnerLife0 replied:
Thanks.

maksm90 commented on src/backend/optimizer/plan/setrefs.c:
```diff
@@ -1380,6 +1380,24 @@ trivial_subqueryscan(SubqueryScan *plan)
+	 * GPDB: Stripping out of SubqueryScan node causing overwriting of
+	 * subplan's flow (see set_subqueryscan_references()). The following check
+	 * restricts such overwriting.
+	 * 
```
Trailing space

InnerLife0 replied:
Thanks.

maksm90 commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -358,3 +358,63 @@ WITH RECURSIVE r1 AS (
+with cte as (
+    insert into with_dml select i, i * 100 from generate_series(1,5) i
+    returning i
+) select count(*) from cte;
```
You might add `where` clause for `select` stmt to verify whether it pushes down to `insert` one.

maksm90 commented:
I'm confused that after second part of patch the following case becomes possible:
```sql
explain (costs off)
with cte as (
    delete from with_dml_dr where i > 0 returning i
) select count(*) from cte join with_dml using(i);
                                         QUERY PLAN                                          
---------------------------------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 2:1  (slice3; segments: 2)
         ->  Aggregate
               ->  Hash Join
                     Hash Cond: (with_dml.i = cte.i)
                     ->  Seq Scan on with_dml
                     ->  Hash
                           ->  Redistribute Motion 1:2  (slice2; segments: 1)
                                 Hash Key: cte.i
                                 ->  Subquery Scan on cte
                                       ->  Explicit Gather Motion 2:1  (slice1; segments: 2)
                                             ->  Delete on with_dml_dr
                                                   ->  Seq Scan on with_dml_dr
                                                         Filter: (i > 0)
 Optimizer: Postgres query optimizer
```
When we join distributed replicated dataset after modification with partitioned table we observe redundant round trip (from one segment to coordinator and back dis-jointly to corresponding segments) of tuples from replicated dataset. Oppositely, if we exchange the table types for join:
```sql
explain (costs off)
with cte as (
    delete from with_dml where i > 0 returning i
) select count(*) from cte join with_dml_dr using(i);
                         QUERY PLAN                          
-------------------------------------------------------------
 Aggregate
   ->  Gather Motion 2:1  (slice1; segments: 2)
         ->  Aggregate
               ->  Hash Join
                     Hash Cond: (with_dml_dr.i = with_dml.i)
                     ->  Seq Scan on with_dml_dr
                     ->  Hash
                           ->  Delete on with_dml
                                 ->  Seq Scan on with_dml
                                       Filter: (i > 0)
 Optimizer: Postgres query optimizer
```
we end up to localized joining.

In general we might achieve localized join for the first case. @InnerLife0 is any way to do this?

maksm90 commented:
UPDATE/DELETE partitioned table to which multiple time refers upper query fails with following assert:
```sql
explain (costs off, verbose) 
with cte as (
    update with_dml set j = j + 1 where i <= 5 
    returning i
) select count(1) from cte where i < (select avg(i) from cte);
FATAL:  Unexpected internal error (cdbmutate.c:1203)
DETAIL:  FailedAssertion("!(segidColIdx > 0 && segidColIdx <= list_length(lefttree->targetlist))", File: "cdbmutate.c", Line: 1203)
HINT:  Process 208406 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.

explain (costs off, verbose) 
with cte as (
    delete from with_dml where i > 0
    returning i
) select count(1) from cte where i < (select avg(i) from cte);
FATAL:  Unexpected internal error (cdbmutate.c:1203)
DETAIL:  FailedAssertion("!(segidColIdx > 0 && segidColIdx <= list_length(lefttree->targetlist))", File: "cdbmutate.c", Line: 1203)
HINT:  Process 208898 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
```
Backtrace for UPDATE case is following:
```
#4  0x000055d900399cef in ExceptionalCondition (conditionName=0x55d900b15808 "!(segidColIdx > 0 && segidColIdx <= list_length(lefttree->targetlist))", errorType=0x55d900b14e4c "FailedAssertion", 
    fileName=0x55d900b14e40 "cdbmutate.c", lineNumber=1203) at assert.c:46
#5  0x000055d900491e8a in make_explicit_motion (lefttree=0x55d9070f5aa0, segidColIdx=0, useExecutorVarFormat=1 '\001') at cdbmutate.c:1203
#6  0x000055d90049149a in apply_motion_mutator (node=0x55d9070f05d8, context=0x7ffcb7beaa10) at cdbmutate.c:938
#7  0x000055d90005ba5a in expression_tree_mutator (node=0x55d9070f0578, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at nodeFuncs.c:2768
#8  0x000055d9004b744f in plan_tree_mutator (node=0x55d9070f0578, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:921
#9  0x000055d900490f9c in apply_motion_mutator (node=0x55d9070f0578, context=0x7ffcb7beaa10) at cdbmutate.c:782
#10 0x000055d9004b463c in plan_tree_mutator (node=0x55d9070f0298, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:193
#11 0x000055d90049106d in apply_motion_mutator (node=0x55d9070f0298, context=0x7ffcb7beaa10) at cdbmutate.c:804
#12 0x000055d9004b74d4 in mutate_plan_fields (newplan=0x55d9070f5500, oldplan=0x55d9070eff20, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:946
#13 0x000055d9004b60cd in plan_tree_mutator (node=0x55d9070eff20, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:607
#14 0x000055d90049106d in apply_motion_mutator (node=0x55d9070eff20, context=0x7ffcb7beaa10) at cdbmutate.c:804
#15 0x000055d9004b74d4 in mutate_plan_fields (newplan=0x55d9070f52d8, oldplan=0x55d9070efcd0, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:946
#16 0x000055d9004b6aa3 in plan_tree_mutator (node=0x55d9070efcd0, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:698
#17 0x000055d90049106d in apply_motion_mutator (node=0x55d9070efcd0, context=0x7ffcb7beaa10) at cdbmutate.c:804
#18 0x000055d9004b74d4 in mutate_plan_fields (newplan=0x55d9070f4f60, oldplan=0x55d9070efa50, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:946
#19 0x000055d9004b60cd in plan_tree_mutator (node=0x55d9070efa50, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:607
#20 0x000055d90049106d in apply_motion_mutator (node=0x55d9070efa50, context=0x7ffcb7beaa10) at cdbmutate.c:804
#21 0x000055d9004b6eb5 in plan_tree_mutator (node=0x55d9070e0df8, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:754
#22 0x000055d900490f63 in apply_motion_mutator (node=0x55d9070e0df8, context=0x7ffcb7beaa10) at cdbmutate.c:774
#23 0x000055d90005ba5a in expression_tree_mutator (node=0x55d9070ef958, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at nodeFuncs.c:2768
#24 0x000055d9004b744f in plan_tree_mutator (node=0x55d9070ef958, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:921
#25 0x000055d900490f9c in apply_motion_mutator (node=0x55d9070ef958, context=0x7ffcb7beaa10) at cdbmutate.c:782
#26 0x000055d9004b751e in mutate_plan_fields (newplan=0x55d9070f1650, oldplan=0x55d9070eb740, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:948
#27 0x000055d9004b60cd in plan_tree_mutator (node=0x55d9070eb740, mutator=0x55d900490e3d <apply_motion_mutator>, context=0x7ffcb7beaa10) at cdbplan.c:607
#28 0x000055d90049106d in apply_motion_mutator (node=0x55d9070eb740, context=0x7ffcb7beaa10) at cdbmutate.c:804
#29 0x000055d900490b56 in apply_motion (root=0x55d9070d4b68, plan=0x55d9070eb740, query=0x55d9070d2c48) at cdbmutate.c:681
#30 0x000055d90048bf01 in cdbparallelize (root=0x55d9070d4b68, plan=0x55d9070eb740, query=0x55d9070d2c48) at cdbllize.c:210
#31 0x000055d90010e66d in standard_planner (parse=0x55d9070d2c48, cursorOptions=0, boundParams=0x0) at planner.c:401
#32 0x000055d90010de59 in planner (parse=0x55d907007d18, cursorOptions=0, boundParams=0x0) at planner.c:200
#33 0x000055d90020db80 in pg_plan_query (querytree=0x55d907007d18, cursorOptions=0, boundParams=0x0) at postgres.c:971
#34 0x000055d8fff0e0ce in ExplainOneQuery (query=0x55d907007d18, into=0x0, es=0x7ffcb7beade0, 
    queryString=0x55d906f2a048 "explain (costs off, verbose) \nwith cte as (\n    update with_dml set j = j + 1 where i <= 5 \n    returning i\n) select count(1) from cte where i < (select avg(i) from cte);", 
    params=0x0) at explain.c:410
```

InnerLife0 commented:
> UPDATE/DELETE partitioned table to which multiple times refers upper query fails with following assert:

The following resolved in new (third) commit.

hughcapet commented:
Am I wrong: this PR is not about fixing an incorrect behavior of SELECTs over _dml_. It is rather about problems in SELECTs over (CTEs with) _RETURNING_ clause, isn't it?

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1235,7 +1236,15 @@ cdbpath_motion_for_join(PlannerInfo *root,
-		if (root->upd_del_replicated_table > 0)
+		if (root->parse->hasModifyingCTE && root->parse->commandType == CMD_SELECT)
+		{
+			/*
+			 * ignore special cases for SELECT over DML, we don't need to make
+			 * inner part of join replicated and add additional motion layers
+			 * later
+			 */
```
As far I understand, we should handle `CdbLocusType_Replicated` in these two blocks. There are asserts for only Bottlenecks and Partitioned locuses in blocks below.
But we never add motions in case of `CdbLocusType_Replicated` type in the current implementation. When is it a problem? E.g. when we join our DML CTE with data available on the master only:

1. Master-only tables:
    ```
    2160=# explain (costs off) with cte as (                                                                                                                                                                
        insert into with_dml_dr select i, i * 100 from generate_series(2658,2664) i
        returning i
    ) select * from cte join pg_class on cte.i = oid;
                         QUERY PLAN                     
    ----------------------------------------------------
     Hash Join
       Hash Cond: ((with_dml_dr.i)::oid = pg_class.oid)
       ->  Insert on with_dml_dr
             ->  Function Scan on generate_series i
       ->  Hash
             ->  Seq Scan on pg_class
     Optimizer: Postgres query optimizer
    (7 rows)
    ```
3. `CdbLocusType_Entry`
    ```
    2160=# explain (costs off) with cte as (
        insert into with_dml_dr select i, i * 100 from generate_series(1,5) i
        returning i
    ) select * from cte left join (select * from with_dml limit 5) lmt on cte.i = lmt.i;
                             QUERY PLAN                         
    ------------------------------------------------------------
     Hash Left Join
       Hash Cond: (with_dml_dr.i = with_dml.i)
       ->  Insert on with_dml_dr
             ->  Function Scan on generate_series i
       ->  Hash
             ->  Limit
                   ->  Gather Motion 3:1  (slice1; segments: 3)
                         ->  Limit
                               ->  Seq Scan on with_dml
     Optimizer: Postgres query optimizer
    (10 rows)
    ```

These plans return nothing to master.

InnerLife0 replied:
Fixed according to our voice talk. Tests added.

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1102,7 +1102,8 @@ cdbpath_motion_for_join(PlannerInfo *root,
 			{
-				Assert(root->upd_del_replicated_table > 0);
+				Assert(root->upd_del_replicated_table > 0 ||
+					   (root->parse->hasModifyingCTE && root->parse->commandType == CMD_SELECT));
```
Can you remind me, did you test case when we join modifying CTE (under 3-numsegments table) results with 2-numsegments replicated table? We decrease numsegments for `other` locus, but execute it on 3 segments. But data for the 2nd table is located only on 2 segments.

InnerLife0 replied:
As far as I remember, we did that test case. Do you want to see it in tests?

InnerLife0 replied:
Comment added.

Commit:
Fixing SELECT over DML queries.
Previously, using of SELECT over DML operations caused various errors.
The first block of errors is related to distributed replicated tables. There was no motion node between SELECT and DML plan nodes. This caused SELECT to work with empty result set, no matter how many rows were processed by DML beneath.
Another error was with joining such result set with any table. There were assertion errors due to incompatibility of CdbLocusType_Replicated with SELECT statements.
The fix adds additional analysis of hasModifyingCTE flag which assumes SELECT query can have DML beneath.

Commit:
The second block of errors is related to quals pushed down from SELECT to DML operation. It's incorrect to push down quals to DML operations, because manipulated dataset can be filtered by wrong upper quals. It also caused indistinct rewrite error.

Commit:
The third block of errors is related to multiple references to CTE in SELECT. It causes duplicated DML operations or mutation errors during apply_motion() of cdbparallelize().
To fix this, we restrict SELECT queries with multiple references to CTE with DML.

Commit:
Fixup for with_dml_dr_seg1 table

hughcapet commented:
I'm lovin' in

```
explain (costs off) with cte as (
insert into with_dml_dr select i, i * 100  from generate_series(1,5) i 
returning i, j) select * from cte union select * from with_dml;
ERROR:  unexpected arguments to set operation (cdbsetop.c:231)
```

```
explain (costs off) with cte as (  
insert into with_dml_dr select i, i * 100 from generate_series(1,5) I
returning I
select distinct count(*) from cte;
ERROR:  unexpected input locus to distinct (planner.c:3026)
```

Commit:
Additional fixes and tests related to replicated tables
